src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs

```diff
@@ -597,18 +597,48 @@ internal bool IsEntryPointCandidate
 
         /// <summary>
         /// Checks if the method has an entry point compatible signature, i.e.
-        /// - the return type is either void or int
+        /// - the return type is either void, int, <see cref="System.Threading.Tasks.Task" />, 
+        /// or <see cref="System.Threading.Tasks.Task{T}" /> Where T is an int.
         /// - has either no parameter or a single parameter of type string[]
         /// </summary>
-        internal bool HasEntryPointSignature()
+        internal bool HasEntryPointSignature(CSharpCompilation compilation)
         {
-            if (this.IsVararg)
+
+            bool IsAsyncMainReturnType(TypeSymbol type)
```

Comment:
Should the new logic be conditional on language version?
If so, please add a test with language version 7.